core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerPool.scala

```diff
@@ -0,0 +1,866 @@
+      val oldRevision = deletionMessage.revision
+      val invocationNamespace = deletionMessage.invocationNamespace
+      val fqn = deletionMessage.action.copy(version = None)
+
+      warmedPool.foreach(warmed => {
+        val proxy = warmed._1
+        val data = warmed._2
+
+        if (data.invocationNamespace == invocationNamespace
+            && data.action.fullyQualifiedName(withVersion = false) == fqn.copy(version = None)
+            && data.revision <= oldRevision) {
+          proxy ! GracefulShutdown
+        }
+      })
+
+      busyPool.foreach(f = busy => {
+        val proxy = busy._1
+        busy._2 match {
+          case warmData: WarmData
+              if warmData.invocationNamespace == invocationNamespace
+                && warmData.action.fullyQualifiedName(withVersion = false) == fqn.copy(version = None)
+                && warmData.revision <= oldRevision =>
+            proxy ! GracefulShutdown
+          case initializedData: InitializedData
+              if initializedData.invocationNamespace == invocationNamespace
+                && initializedData.action.fullyQualifiedName(withVersion = false) == fqn.copy(version = None) =>
+            proxy ! GracefulShutdown
+          case _ => // Other actions are ignored.
+        }
+      })
+
+    case ReadyToWork(data) =>
+      prewarmStartingPool = prewarmStartingPool - sender()
+      prewarmedPool = prewarmedPool + (sender() -> data)
+      prewarmCreateFailedCount = prewarmCreateFailedCount - ((data.kind, data.memoryLimit))
+
+    // Container is initialized
+    case Initialized(data) =>
+      busyPool = busyPool + (sender() -> data)
+      inProgressPool = inProgressPool - sender()
+      // container init completed, send creationAck(success) to scheduler
+      creationMessages.remove(sender()).foreach { msg =>
+        val ack = ContainerCreationAckMessage(
+          msg.transid,
+          msg.creationId,
+          msg.invocationNamespace,
+          msg.action,
+          msg.revision,
+          msg.whiskActionMetaData,
+          instance,
+          msg.schedulerHost,
+          msg.rpcPort,
+          msg.retryCount)
+        sendAckToScheduler(msg.rootSchedulerIndex, ack)
+      }
+
+    case Resumed(data) =>
+      busyPool = busyPool + (sender() -> data)
+      inProgressPool = inProgressPool - sender()
+      // container init completed, send creationAck(success) to scheduler
+      creationMessages.remove(sender()).foreach { msg =>
+        val ack = ContainerCreationAckMessage(
+          msg.transid,
+          msg.creationId,
+          msg.invocationNamespace,
+          msg.action,
+          msg.revision,
+          msg.whiskActionMetaData,
+          instance,
+          msg.schedulerHost,
+          msg.rpcPort,
+          msg.retryCount)
+        sendAckToScheduler(msg.rootSchedulerIndex, ack)
+      }
+
+    // if warmed containers is failed to resume, we should try to use other container or create a new one
```

Comment:
should we attempt to remove the container that failed to remove as well?

Reply:
Yes, i think currently use `info` is better, after all prs of the scheduler are merged and become stable, i think we can submit a separate pr to change all relative log level.